test/e2e/network/endpointslice.go

```diff
@@ -299,6 +303,146 @@ var _ = SIGDescribe("EndpointSlice", func() {
 		deleteEndpointSlices(cs, f.Namespace.Name, svc2)
 		expectEndpointsAndSlices(cs, f.Namespace.Name, svc2, []*v1.Pod{pod1, pod2}, 2, 2, true)
 	})
+
+	// Ensure that garbage collection works well when a Service is deleted and
+	// its EndpointSlices are being rapidly updated.
+	// For more context: https://github.com/kubernetes/kubernetes/issues/91287
+	ginkgo.It("should not get in the way of Service owner garbage collection", func() {
+		sharedLabels := map[string]string{"test-garbage-collection": "for-service"}
+
+		deployment, err := cs.AppsV1().Deployments(f.Namespace.Name).Create(context.TODO(), &appsv1.Deployment{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:   "test-garbage-collection",
+				Labels: sharedLabels,
+			},
+			Spec: appsv1.DeploymentSpec{
+				Selector: &metav1.LabelSelector{MatchLabels: sharedLabels},
+				Replicas: utilpointer.Int32Ptr(1),
+				Template: v1.PodTemplateSpec{
+					ObjectMeta: metav1.ObjectMeta{
+						Labels: sharedLabels,
+					},
+					Spec: v1.PodSpec{
+						Containers: []v1.Container{
+							{
+								Name:  "container1",
+								Image: imageutils.GetE2EImage(imageutils.Nginx),
+								Ports: []v1.ContainerPort{{
+									Name:          "example-name",
+									ContainerPort: int32(3000),
+								}},
+							},
+						},
+					},
+				},
+			},
+		}, metav1.CreateOptions{})
+		framework.ExpectNoError(err, "Error creating %s/%s Deployment", f.Namespace.Name, deployment.Name)
+
+		ownerRef := metav1.NewControllerRef(deployment, schema.GroupVersionKind{Group: "apps", Version: "v1", Kind: "Deployment"})
+		svc := createServiceReportErr(cs, f.Namespace.Name, &v1.Service{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:            "example-int-port",
+				Labels:          sharedLabels,
+				OwnerReferences: []metav1.OwnerReference{*ownerRef},
+			},
+			Spec: v1.ServiceSpec{
+				// No need to wait for pods to be ready for this test
+				PublishNotReadyAddresses: true,
+				Selector:                 sharedLabels,
+				Ports: []v1.ServicePort{{
+					Name:       "example",
+					Port:       80,
+					TargetPort: intstr.FromInt(3000),
+					Protocol:   v1.ProtocolTCP,
+				}},
+			},
+		})
+		lSelector := discoveryv1beta1.LabelServiceName + "=" + svc.Name
+		maxResourceVersion := waitForNumEndpointSlices(cs, f.Namespace.Name, lSelector, 1)
+
+		deletePropagationForeground := metav1.DeletePropagationForeground
+		err = cs.AppsV1().Deployments(f.Namespace.Name).Delete(context.TODO(), deployment.Name, metav1.DeleteOptions{
+			PropagationPolicy: &deletePropagationForeground,
+		})
+		framework.ExpectNoError(err, "Error deleting %s/%s Deployment", f.Namespace.Name, deployment.Name)
+
+		// There's some delay between when a Service is deleted and the
+		// EndpointSlice informer is updated with that information. This
+		// allows for up to 2 EndpointSlice creations within 15 seconds of
```

Comment:
is this time deterministic?
I'm concerned about possible flakes because of being too optimistic with the timings :sweat_smile:

Reply:
I made this even more lenient so it should never fail, and if it did, I think it would represent a real issue. With that said, I'm open to removing it altogether.